### PR TITLE
Clean up ride code

### LINKED ISSUE #19451: [7aaaa2] openrct2.exe: std::_Narrow_char_traits<char,int>::length, FormatArgument, FormatArgumentAny, OpenRCT2::FormatStringA...
<table>
    <tr>
        <td></td>
        <td><b><a href="https://openrct2.sp.backtrace.io/dashboard/openrct2/project/openrct2/group/7aaaa24b58f139af357e4387d1e2fb15929df7c501cd788b7170dbd5cac30316">New error: [openrct2]</a></b><br /><b>1</b> occurrence since (Fri Nov 15 2019 11:27:39)<br /></td>
    </tr>
</table><table>
    <tr><td><strong>Application</strong></td><td>openrct2.exe</td></tr>
<tr>
    <td><strong>Attributes</strong></td>
    <td>
<strong>_rxid</strong> &mdash; 1
        <ul>
                <li><code>c5000000-2e3a-a309-0000-000000000000</code> &mdash; 1</li></ul><br>
<strong>attachment_config.ini</strong> &mdash; 1
        <ul>
                <li><code>C:\Users\eblue\Documents\OpenRCT2\crash\26bd8a2c-acfa-484a-899e-48d66ea06f39.ini</code> &mdash; 1</li></ul><br>
<strong>assert_failure</strong> &mdash; 1
        <ul>
                <li><code>*</code> &mdash; 1</li></ul><br>
<strong>fault.address</strong> &mdash; 1
        <ul>
                <li><code>2759558165872</code> &mdash; 1</li></ul><br>
<strong>version</strong> &mdash; 1
        <ul>
                <li><code>OpenRCT2, v0.4.3-288-g6f1a1f2 (6f1a1f2 on develop) provided by GitHub</code> &mdash; 1</li></ul><br>
<strong>error.message</strong> &mdash; 1
        <ul>
                <li><code>EXCEPTION_ACCESS_VIOLATION_READ</code> &mdash; 1</li></ul><br>
<strong>commit</strong> &mdash; 1
        <ul>
                <li><code>6f1a1f2</code> &mdash; 1</li></ul><br></td>
    </tr>
    <tr>
        <td><strong>Classifiers</strong></td>
        <td>
            <strong>invalid-read</strong><br></td>
    </tr>
    <tr>
        <td><strong>Frames</strong></td>
        <td>
            <code>
std::_Narrow_char_traits&lt;char,int&gt;::length
FormatArgument
FormatArgumentAny
OpenRCT2::FormatStringAny
FormatStringAny
OpenRCT2::FormatStringLegacy
std::basic_string&lt;char,std::char_traits&lt;char&gt;,std::allocator&lt;char&gt; &gt;::{ctor}
GuestListWindow::DrawScrollSummarised
GuestListWindow::OnScrollDraw
std::allocator&lt;char&gt;::allocate
WidgetScrollDraw
WidgetDraw
WindowDrawWidgets
Window::DrawWidgets
GuestListWindow::OnDraw
WindowEventPaintCall
WindowDrawSingle
WindowDrawCore
WindowDraw
invoke
std::_Invoker_ret&lt;void&gt;::_Call
std::_Func_impl_no_alloc&lt;`WindowDrawAll&#39;::`2&#39;::&lt;lambda_1&gt;,void,WindowBase *&gt;::_Do_call
WindowVisitEach(std::function&lt;void __cdecl
WindowDrawAll
OpenRCT2::Drawing::X8DrawingEngine::DrawDirtyBlocks
OpenRCT2::Drawing::X8DrawingEngine::DrawAllDirtyBlocks
OpenRCT2::Drawing::X8DrawingEngine::PaintWindows
OpenRCT2::Paint::Painter::Paint
OpenRCT2::Context::Draw
OpenRCT2::Context::RunVariableFrame
OpenRCT2::Context::RunFrame
OpenRCT2::Context::RunGameLoop
OpenRCT2::Context::Launch
OpenRCT2::Context::RunOpenRCT2
NormalisedMain
wmain
invoke_main</code>
        </td>
    </tr></table>

### PR DESCRIPTION
**Pull request by ZehMatt**
This PR changes a couple of things. 
1. Ride storage is now a an fixed array, instead of fragmenting memory this is one single flat array.
2. Added a new function for deleting Rides, this makes it easier to keep track of what rides are used.
3. Rides will be now considered as free if their id is RideId::Null, RIDE_TYPE_NULL is an odd way to communicate that, RIDE_TYPE_NULL is also used for other things.
4. Renamed GetOrAllocateRide to RideAllocateAtIndex, I've added asserts that ensures the slot is free for debug builds.
5. Fixed iterations of deleted rides, if you had 1000 rides and deleted all of them then it would forever skip empty 1000 rides during iterations, its still possible to have gaps in the iteration but it will now adjust the maximum iteration size when the last ride in the list is deleted.
6. Track previews no longer share the ride slot 0, there is now a dedicated instance for this purpose, now there is no question anymore if the first ride with id 0 will be affected in any way, it was primarily safe due to the ghost flag in the game actions, now it can in theory mutate the ride without consequences.
7. Happens to Close #19451, the issue was due to Rides being stored in a vector, the vector may grow the buffer and all pointers stored in Formatters will be invalid, this is primarily due to the custom_name field in Ride which is std::string and during vector growth the string would be copied, the easiest way to trigger is this to have the guest list open and have guests with custom names and then open the track designs, since they allocate a new ride this will grow the vector in specific conditions, creating new rides and the guest list will also do the trick.

There is still a lot of code that can be cleaned up, most things should be probably be placed in RideManager, currently it acts only as an way to obtain iterators, the RideManager should take care of things like creation, deletion, updating rides, but this is for another PR.

### CODE REVIEW

File: src/openrct2/ride/Ride.cpp [hunk 1]
```diff
@@ -95,7 +95,14 @@ static constexpr const int32_t RideInspectionInterval[] = {
     10, 20, 30, 45, 60, 120, 0, 0,
 };
 
-static std::vector<Ride> _rides;
+// Ride storage for all the rides in the park, rides with RideId::Null are considered free.
+static std::array<Ride, OpenRCT2::Limits::MaxRidesInPark> _rides{};
+
+// This is not the real size of rides in use, rather the highest slot used + 1.
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:101]: then perhaps name it a bit more obvious then.

File: src/openrct2/ride/Ride.cpp [hunk 2]
```diff
@@ -904,8 +936,12 @@ bool Ride::SupportsStatus(RideStatus s) const
  */
 void RideInitAll()
 {
-    _rides.clear();
-    _rides.shrink_to_fit();
+    std::for_each(std::begin(_rides), std::end(_rides), [](auto& ride) {
+        ride.id = RideId::GetNull();
+        ride.type = RIDE_TYPE_NULL;
+        ride.custom_name = {};
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:942]: should we also do measurement = {}? The fact this is the same as RideDelete makes me think it should be extracted out to a method.

Comment (ZehMatt) [reply src/openrct2/ride/Ride.cpp]: Agree and done.

File: src/openrct2/ride/Ride.cpp [hunk 3]
```diff
@@ -98,8 +98,8 @@ static constexpr const int32_t RideInspectionInterval[] = {
 // Ride storage for all the rides in the park, rides with RideId::Null are considered free.
 static std::array<Ride, OpenRCT2::Limits::MaxRidesInPark> _rides{};
 
-// This is not the real size of rides in use, rather the highest slot used + 1.
-static size_t _maxRideSize = 0;
+// This is the highest used index + 1 of the _rides array.
+static size_t _maxHighestUsedSlot = 0;
```
Comment (Broxzier) [inline src/openrct2/ride/Ride.cpp:102]: 'Max' suggests that it's an absolute maximum possible, but it's pretty much the end of use the used range if I understand this correctly. Perhaps a name like `_endOfUsedRange` or something similar better explains what this is exactly.

Comment (ZehMatt) [reply src/openrct2/ride/Ride.cpp]: Changed

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**